protocol-units/bridge/chains/ethereum/src/lib.rs

```diff
@@ -91,17 +105,44 @@ impl EthClient {
 			.wallet(EthereumWallet::from(config.signer_private_key.clone()))
 			.on_builtin(config.rpc_url.as_str())
 			.await?;
-		let ws = WsConnect::new(config.ws_url.clone());
-		let ws_provider = ProviderBuilder::new().on_ws(ws).await?;
+
+		// let ws = WsConnect::new(ws_url);
+		// println!("ws {:?}", ws);
+		// let ws_provider = ProviderBuilder::new().on_ws(ws).await?;
+		// println!("ws_provider {:?}", ws_provider);
 		Ok(EthClient {
 			rpc_provider,
-			ws_provider,
-			initiator_contract: config.initiator_contract,
-			counterparty_contract: config.counterparty_contract,
+			rpc_port: 8545,
+			ws_provider: None,
+			initiator_contract: None,
+			counterparty_contract: None,
 			config,
 		})
 	}
 
+	pub fn set_initiator_contract(&mut self, contract: InitiatorContract) {
+		self.initiator_contract = Some(contract);
+	}
+
+	pub fn set_counterparty_contract(&mut self, contract: CounterpartyContract) {
+		self.counterparty_contract = Some(contract);
+	}
+
+	pub async fn initialize_initiator_contract(
+		&self,
+		weth: EthAddress,
+		owner: EthAddress,
+	) -> Result<(), anyhow::Error> {
+		let contract =
+			AtomicBridgeInitiator::new(self.initiator_contract_address()?, &self.rpc_provider);
```

Comment:
Seems like you are initializing the implementation contract here. It should impede you from using it as a functional contract but there might be something you are trying to call that isn't included in the implementation contract, leading to a fail to estimate gas error

Reply:
Yeah I tried before storing the contract instance in the Client. But I get the same error. This doesn't create a new contract, but instead creates a wrapped around the onchain instance. So state should persist. This is also the same pattern in MCR settlement.

Reply:
This call does internally call weth, but that the WETH interface is imported inside `AtomicBridgeInitiator`, and so its calls are encoded into `AtomicBridgeInitiator` abi `bytecode` field, so it should be fine.

Reply:
I've just tested this again and updated it to use the method I created that fetches the previously initialised contract instance:
```rust
let contract = self.initiator_contract().expect("Initiator contract not set");
		let recipient_bytes: [u8; 32] =
			recipient_address.0.try_into().expect("Recipient address must be 32 bytes");
		let call = contract
			.initiateBridgeTransfer(
				U256::from(0), // For now a 0 WETH amount
				FixedBytes(recipient_bytes),
				FixedBytes(hash_lock.0),
				U256::from(time_lock.0),
			)
```
So fetching the same initialised contract. It fails with the same error.